packages/polling-controller/src/PollingController.ts

```diff
@@ -142,6 +142,8 @@ function PollingControllerMixin<TBase extends Constructor>(Base: TBase) {
       if (this.#intervalIds[key]) {
         clearTimeout(this.#intervalIds[key]);
         delete this.#intervalIds[key];
+      } else {
+        this._executePoll(networkClientId, options);
```

Comment:
we don't await this. Should we go one step further and ensure executePoll is always executed async, .i.e. we never wait for it to return?

Reply:
I believe we want to await these calls because otherwise they might return out of order which could create inaccurate state?

Reply:
waiting for the call on this line to complete will block the return of the polling token in `startPollingByNetworkClientId()`

Reply:
🤔 I'm not sure what to do here. @shanejonas or @mcmire thoughts?

Reply:
I agree with @jiexi that `startPollingByNetworkClientId` can't await the call to `this.#poll`, but that doesn't mean we can't `await` this line here, right?

Then again, I'm not sure what you mean by this here @adonesky1, can you go into more detail?

> I believe we want to await these calls because otherwise they might return out of order which could create inaccurate state?

Reply:
Like so lets say the polling interval is 3 seconds and there is a series of calls to `this._executePoll` and say for whatever reason the network request associated with - in this case - the first call hung for a while and returned after 10 seconds, meanwhile since we're not awaiting here, we've gone ahead to the second and third calls (which responded immediately) then the third call would actually resolve last but with outdated info meaning we will update state with very outdated info? Maybe I'm missing something though.

Reply:
Might be nicer to start off the polling with `0` timer and let the async/await happen there before #poll is called.

replacing 151-158 with something like:

```
      this.#intervalIds[key] = setTimeout(async () => {
        try {
          await this._executePoll(networkClientId, options);
        } catch (error) {
          console.error(error);
        }
        this.#poll(networkClientId, {...options, first: false});
      }, options.first ? 0 : this.#intervalLength);
```

Reply:
not saying `options.first` is the right approach for the parameters here. maybe we add another param, just a quick example that got my thoughts across.

Reply:
what about a `        void new Promise(async() => this._executePoll(networkClientId, options))`

Reply:
> what about a `void new Promise(() => this._executePoll(networkClientId, options))`

we need to `await _executePoll` like it does inside `setTimeout` or else we would be in the same place we were with `setInterval` not dealing with execution times.

Might as well use what we have, set it to 0 first time around.

Reply:
> Like so lets say the polling interval is 3 seconds and there is a series of calls to this._executePoll and say for whatever reason the network request associated with - in this case - the first call hung for a while and returned after 10 seconds, meanwhile since we're not awaiting here, we've gone ahead to the second and third calls (which responded immediately) then the third call would actually resolve last but with outdated info meaning we will update state with very outdated info? Maybe I'm missing something though.

If I'm understanding you correctly, you're saying that we should only schedule another poll once we have a poll that completes?

Why can't we: 

- `await` the call to `_executePoll`
- _not_ `await` the call to `this.#poll` in `startPollingByNetworkClientId`

?

Reply:
how about this then

```
    async #poll(networkClientId: NetworkClientId, options: Json) {
      const key = getKey(networkClientId, options);
      if (this.#intervalIds[key]) {
        clearTimeout(this.#intervalIds[key]);
        delete this.#intervalIds[key];
      } else {
        await this.#safelyExecutePoll(networkClientId, options)
      }
      // setTimeout is not `await`ing this async function, which is expected
      // We're just using async here for improved stack traces
      // eslint-disable-next-line @typescript-eslint/no-misused-promises
      this.#intervalIds[key] = setTimeout(async () => {
        await this.#safelyExecutePoll(networkClientId, options)
        this.#poll(networkClientId, options);
      }, this.#intervalLength);
    }

    async #safelyExecutePoll(networkClientId: NetworkClientId, options: Json) {
      try {
        await this._executePoll(networkClientId, options)
      } catch (error) {
        console.error(error)
      }
    }
```

Reply:
I fixed my suggestion to rely on `this.#intervalIds[key]` as well, simplifies it from being an argument, and leaves `#poll` as sync:

```
      this.#intervalIds[key] = setTimeout(async () => {
        try {
          await this._executePoll(networkClientId, options);
        } catch (error) {
          console.error(error);
        }
        this.#poll(networkClientId, {...options, first: false});
      }, this.#intervalIds[key] ? 0 : this.#intervalLength);
      
```

Reply:
`this.#intervalIds[key]` will be falsey at this point because it's always cleared above this snippet of code

Maybe something like this?
```
    #poll(networkClientId: NetworkClientId, options: Json) {
      const key = getKey(networkClientId, options);
      const interval = this.#intervalIds[key]
      if (interval) {
        clearTimeout(interval);
        delete this.#intervalIds[key];
      }
      // setTimeout is not `await`ing this async function, which is expected
      // We're just using async here for improved stack traces
      // eslint-disable-next-line @typescript-eslint/no-misused-promises
      this.#intervalIds[key] = setTimeout(async () => {
        try {
          await this._executePoll(networkClientId, options);
        } catch (error) {
          console.error(error);
        }
        this.#poll(networkClientId, options);
      }, interval ? this.#intervalLength : 0);
    }
```

Reply:
done https://github.com/MetaMask/core/pull/1874/commits/b5dc78ac703fbc2d7695975551078a2949a8446f